src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlConnectionDetails.cs

```diff
@@ -0,0 +1,180 @@
+                        InstanceName = instanceName,
+                        Port = null,
+                    };
+                }
+            }
+
+            return new SqlConnectionDetails
+            {
+                ServerHostName = serverHostName,
+                ServerIpAddress = serverIpAddress,
+                InstanceName = null,
+                Port = null,
+            };
+        }
+
+        string? port;
+        if (match.Groups[4].Length > 0)
+        {
+            instanceName = match.Groups[3].Value;
+            port = match.Groups[4].Value;
+            if (port == "1433")
+            {
+                port = null;
+            }
+        }
+        else if (int.TryParse(match.Groups[3].Value, out int parsedPort))
+        {
+            port = parsedPort == 1433 ? null : match.Groups[3].Value;
+            instanceName = null;
+        }
+        else
+        {
+            instanceName = match.Groups[3].Value;
+
+            if (string.IsNullOrEmpty(instanceName))
+            {
+                instanceName = null;
+            }
+
+            port = null;
+        }
+
+        return new SqlConnectionDetails
+        {
+            ServerHostName = serverHostName,
+            ServerIpAddress = serverIpAddress,
+            InstanceName = instanceName,
+            Port = port,
+        };
+    }
+
+    internal static SqlConnectionDetails GetOrAddCached(string dataSource)
+    {
+        if (!ConnectionDetailCache.TryGetValue(dataSource, out SqlConnectionDetails? connectionDetails))
+        {
+            connectionDetails = ParseDataSource(dataSource);
+            ConnectionDetailCache.TryAdd(dataSource, connectionDetails);
+        }
+
+        return connectionDetails;
+    }
+
+    internal static void AddConnectionLevelDetailsToActivity(SqlClientTraceInstrumentationOptions options, string dataSource, Activity sqlActivity)
+    {
+        if (!options.EnableConnectionLevelAttributes)
+        {
+            sqlActivity.SetTag(SemanticConventions.AttributePeerService, dataSource);
+        }
+        else
+        {
+            var connectionDetails = GetOrAddCached(dataSource);
+
+            if (!string.IsNullOrEmpty(connectionDetails.InstanceName))
+            {
+                sqlActivity.SetTag(SemanticConventions.AttributeDbMsSqlInstanceName, connectionDetails.InstanceName);
+            }
+
+            if (!string.IsNullOrEmpty(connectionDetails.ServerHostName))
+            {
+                sqlActivity.SetTag(SemanticConventions.AttributeServerAddress, connectionDetails.ServerHostName);
+            }
+            else
+            {
+                sqlActivity.SetTag(SemanticConventions.AttributeServerSocketAddress, connectionDetails.ServerIpAddress);
+            }
+
+            if (!string.IsNullOrEmpty(connectionDetails.Port))
+            {
+                // TODO: Should we continue to emit this if the default port (1433) is being used?
```

Comment:
No, it should not be set if it is the default. See the following footnote from the [conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-metrics.md#metric-dbclientoperationduration)

> [11]: If using a port other than the default port for this DBMS and if server.address is set.